File: CSES/Labyrinth.cpp
#include <bits/stdc++.h>
using namespace std;

#define fast ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);

const vector<pair<int, int>> dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
const vector<char> dir_chars = {'U', 'D', 'L', 'R'};
int n, m;
vector<vector<char>> grid;
int path[int(2e6)];

int f(int x, int y) {
    return x * 1000 + y;
}

pair<int, int> bfs(int a, int b) {
    queue<int> q;
    q.push(f(a, b));
    grid[a][b] = '#';
    while (!q.empty())
    {
        int i = q.front() / 1000, j = q.front() % 1000;
        q.pop();
        for (int d = 0; d < 4; ++d)
        {
            int x = i + dirs[d].first, y = j + dirs[d].second;
            if (x >= 0 && x < n && y >= 0 && y < m && grid[x][y] != '#') {
                path[f(x, y)] = d;
                if (grid[x][y] == 'B')
                    return {x, y};
                grid[x][y] = '#';
                q.push(f(x, y));
            }
        }
    }
    return {-1, -1};
}

int main() {
    fast;
    cin >> n >> m;
    grid.resize(n, vector<char>(m));
    int start_x = -1, start_y = -1;
    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++) {
            cin >> grid[i][j];
            if (grid[i][j] == 'A')
                start_x = i, start_y = j;
        }

    if (start_x == -1) {
        cout << "NO\n";
        return 0;
    }
    
    auto [x, y] = bfs(start_x, start_y);
    if (x == -1) {
        cout << "NO\n";
        return 0;
    }
    
    string ans = "";
    while (x != start_x || y != start_y) {
        int d = path[f(x, y)];
        ans += dir_chars[d];
        x -= dirs[d].first;
        y -= dirs[d].second;
    }
    
    reverse(ans.begin(), ans.end());
    cout << "YES\n" << ans.size() << "\n" << ans << endl;
    return 0;
}

File: CSES/Python/Labyrinth.py
from queue import Queue

dirs = [(-1, 0, 'U'), (1, 0, 'D'), (0, -1, 'L'), (0, 1, 'R')]
n, m = map(int, input().split())
grid = [list(input()) for _ in range(n)]
path = dict()

def f(x, y):
    return x * 1000 + y

def bfs(a, b):
    q = Queue()
    q.put(f(a, b))
    while not q.empty():
        idx = q.get()
        i, j = idx // 1000, idx % 1000
        if grid[i][j] == 'B':
            return i, j
        grid[i][j] = '#'
        for dx, dy, di in dirs:
            x, y = i + dx, j + dy
            if 0 <= x < n and 0 <= y < m and grid[x][y] != '#':
                path[f(x, y)] = di
                q.put(f(x, y))
    return -1, -1

for i in range(n):
    for j in range(m):
        if grid[i][j] == 'A':
            x, y = bfs(i, j)
            if x != -1:
                ans = ""
                while not (x == i and j == y):
                    ch = path[f(x, y)]
                    ans += ch
                    for dx, dy, di in dirs:
                        if di == ch:
                            x -= dx
                            y -= dy
                            break
                print(f'YES\n{len(ans)}\n{ans[::-1]}')
            else:
                print("NO")
            break